### PR TITLE
Feature/agent send channel preference report dummy

### PR DESCRIPTION
**Pull request by morantr**
Add send of channel preference report dummy cmdu back to the controller,
upon receiving channel preference query.

#74

### CODE REVIEW

**Comment by tomereli** [comment]
I think you meant to create a draft pull request, doesn't look ready to review - is it?
You have a merge commit.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 1]
```diff
@@ -450,6 +450,15 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
...
 
     auto cmdu_message_type      = cmdu_rx.getMessageType();
+    auto rx_cmdu_message_header = cmdu_rx.getCmduHeader();
+
+    if (!rx_cmdu_message_header) {
+        LOG(ERROR) << "Failed getting cmdu header!";
+        return false;
+    }
+
+    auto rx_mid = rx_cmdu_message_header->message_id();
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: No, this won't work. remove this commit and rebase - for more information see https://github.com/prplfoundation/prplMesh/pull/101

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 2]
```diff
@@ -434,8 +434,6 @@ bool slave_thread::handle_cmdu(Socket *sd, ieee1905_1::CmduMessageRx &cmdu_rx)
 bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
                                                           ieee1905_1::CmduMessageRx &cmdu_rx)
 {
-    auto cmdu_message_type = cmdu_rx.getMessageType();
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: It is also irrelevant for this PR. Such changes should not be part of a feature unless they are needed - in this case this is not required, please remove this commit.

File: common/beerocks/bcl/include/beerocks/bcl/network/network_utils.h [hunk 3]
```diff
@@ -20,6 +20,8 @@
 #include <netinet/ether.h>
 #include <netinet/ip.h> // IP_MAXPACKET (which is 65535)
 
+#include <tlvf/common/sMacAddress.h>
```
Comment (tomereli) [inline common/beerocks/bcl/include/beerocks/bcl/network/network_utils.h]: This commit is also not needed. What is needed is to use what you have even though it is `mac.mac` and add a TODO and a task to unify sMacAddress to a single struct. Please remove this commit.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 4]
```diff
@@ -22,6 +22,8 @@
 #include <beerocks/tlvf/beerocks_message_monitor.h>
 #include <beerocks/tlvf/beerocks_message_platform.h>
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp:24]: remove line break

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 5]
```diff
@@ -460,13 +462,68 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
...
 
     switch (cmdu_message_type) {
-    /* For Example:
+    case ieee1905_1::eMessageType::CHANNEL_PREFERENCE_QUERY_MESSAGE: {
+
+        LOG(INFO) << "received CHANNEL_PREFERENCE_QUERY_MESSAGE";
+
+        // build channel preference report
+        auto cmdu_tx_header =
+            cmdu_tx.create(rx_mid, ieee1905_1::eMessageType::CHANNEL_PREFERENCE_REPORT_MESSAGE);
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: rebase and use cmdu_rx.getMessageId() directly, don't use another variable.

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp [hunk 6]
```diff
@@ -52,6 +52,7 @@ bool master_mrouter::init()
     return (bus_subscribe(std::vector<ieee1905_1::eMessageType>{
         ieee1905_1::eMessageType::VENDOR_SPECIFIC_MESSAGE,
         ieee1905_1::eMessageType::AP_AUTOCONFIGURATION_SEARCH_MESSAGE,
+        ieee1905_1::eMessageType::CHANNEL_PREFERENCE_REPORT_MESSAGE,
```
Comment (tomereli) [inline controller/src/beerocks/master/beerocks_master_mrouter.cpp:56]: please squash this to the previous commit.

**Review by tomereli — CHANGES_REQUESTED** [review]
Last merge commit - what is it? we don't do merge commits.

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 1 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: fixed

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 2 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: fixed

File: common/beerocks/bcl/include/beerocks/bcl/network/network_utils.h (hunk 3 above)
Comment (morantr) [reply common/beerocks/bcl/include/beerocks/bcl/network/network_utils.h]: But I need function that converts mac string to mac of 1905.1.
what is your proposal instead?

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 5 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: This is the first message of 1905.1 we handle on son_slave, and it is more convenient to have a copy of it which can later be used by every message handler.

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp (hunk 6 above)
Comment (morantr) [reply controller/src/beerocks/master/beerocks_master_mrouter.cpp:56]: why? I try to separate commits to atomic units.
I did such a thing in my last PR that merged, and you accepted it then, so why no now?

File: common/beerocks/bcl/include/beerocks/bcl/network/network_utils.h (hunk 3 above)
Comment (tomereli) [reply common/beerocks/bcl/include/beerocks/bcl/network/network_utils.h]: unify sMacAddr to use tlvf - take what I did and do it externally (the operators)

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 5 above)
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: ok

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 7]
```diff
@@ -454,6 +454,8 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
     master_last_seen   = std::chrono::steady_clock::now();
     keep_alive_retries = 0;
 
+    auto rx_mid = cmdu_rx.getMessageId();
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: This doesn't deserve its own commit, do this when you use it.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 8]
```diff
@@ -457,13 +458,68 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
...
+    case ieee1905_1::eMessageType::CHANNEL_PREFERENCE_QUERY_MESSAGE: {
 
-        case ieee1905_1::eMessageType::TOPOLOGY_DISCOVERY_MESSAGE :
-        {
-            break;
+        LOG(INFO) << "received CHANNEL_PREFERENCE_QUERY_MESSAGE";
+
+        // build channel preference report
+        auto cmdu_tx_header =
+            cmdu_tx.create(rx_mid, ieee1905_1::eMessageType::CHANNEL_PREFERENCE_REPORT_MESSAGE);
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: don't call it rx_mid, what's wrong with mid?

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 9]
```diff
@@ -457,13 +458,68 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
...
+        if (!op_class_channels) {
+            LOG(ERROR) << "create_operating_classes_list() has failed!";
+            return false;
+        }
+
+        // Fill operating class object
+        op_class_channels->operating_class() = 80; // random operating class for test purpose
+
+        // Fill up channels list in operating class object, with test values
+        for (uint8_t ch = 36; ch < 50; ch += 2) {
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: This should be using BWL - we have now acs_report also in dummy mode, so use it.

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp (hunk 6 above)
Comment (tomereli) [reply controller/src/beerocks/master/beerocks_master_mrouter.cpp:56]: See here - https://github.com/prplfoundation/prplMesh/pull/66#issuecomment-508678164

Comment (tomereli) [reply controller/src/beerocks/master/beerocks_master_mrouter.cpp:56]: and squash it

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 9 above)
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: See here - https://github.com/prplfoundation/prplMesh/commit/cfcc9b21be1eef6dc4b667ffffb3d5e1ebd07688

**Review by tomereli — CHANGES_REQUESTED** [review]

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 9 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: I understood from the task that what is needed it is create ping pong of dummy messages between the controller and the son_slave, only for certification purpose of the controller alone, without actually drill down the messages to the ap_manager/dwpal_hal.

File: common/beerocks/bcl/include/beerocks/bcl/network/network_utils.h (hunk 3 above)
Comment (morantr) [reply common/beerocks/bcl/include/beerocks/bcl/network/network_utils.h]: fixed

File: controller/src/beerocks/master/beerocks_master_mrouter.cpp (hunk 6 above)
Comment (morantr) [reply controller/src/beerocks/master/beerocks_master_mrouter.cpp:56]: fixed

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 7 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: fixed

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 8 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: fixed

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 10]
```diff
@@ -454,14 +455,71 @@ bool slave_thread::handle_cmdu_control_ieee1905_1_message(Socket *sd,
     master_last_seen   = std::chrono::steady_clock::now();
     keep_alive_retries = 0;
 
+    auto mid = cmdu_rx.getMessageId();
+
     switch (cmdu_message_type) {
-    /* For Example:
+    case ieee1905_1::eMessageType::CHANNEL_PREFERENCE_QUERY_MESSAGE: {
```
Comment (tomereli) [inline agent/src/beerocks/slave/son_slave_thread.cpp:459]: Move this to a separate function

**Review by tomereli — CHANGES_REQUESTED** [review]

File: agent/src/beerocks/slave/son_slave_thread.cpp (hunk 10 above)
Comment (morantr) [reply agent/src/beerocks/slave/son_slave_thread.cpp:459]: done

**Review by tomereli — APPROVED** [review]
